crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,738 @@
+        for q in second_qargs {
+            if !qarg.contains_key(q) {
+                qarg.insert(q, num_qubits);
+                num_qubits += 1;
+            }
+        }
+
+        let first_qarg: Vec<_> = first_qargs.iter().map(|q| *qarg.get(q).unwrap()).collect();
+        let second_qarg: Vec<_> = second_qargs.iter().map(|q| *qarg.get(q).unwrap()).collect();
+
+        assert!(
+            first_qarg.len() <= second_qarg.len(),
+            "first instructions must have at most as many qubits as the second instruction"
+        );
+
+        let first_op = first_instr.op();
+        let second_op = second_instr.op();
+
+        let first_mat = match first_op.matrix(&first_instr.params) {
+            Some(mat) => mat,
+            None => match first_op {
+                PyGateType(gate) => match get_op(&gate.gate) {
+                    Some(x) => x,
+                    _ => return false,
+                },
+                PyOperationType(operation) => match get_op(&operation.operation) {
+                    Some(x) => x,
+                    _ => return false,
+                },
+                _ => return false,
+            },
+        };
+
+        let second_mat = match second_op.matrix(&second_instr.params) {
+            Some(mat) => mat,
+            None => match second_op {
+                PyGateType(gate) => match get_op(&gate.gate) {
+                    Some(x) => x,
+                    _ => return false,
+                },
+                PyOperationType(operation) => match get_op(&operation.operation) {
+                    Some(x) => x,
+                    _ => return false,
+                },
+                _ => return false,
+            },
+        };
+
+        if first_qarg == second_qarg {
+            abs_diff_eq!(
+                second_mat.dot(&first_mat),
+                first_mat.dot(&second_mat),
+                epsilon = 1e-8
+            )
+        } else {
+            let extra_qarg2 = num_qubits - first_qarg.len();
+            let first_mat = if extra_qarg2 > 0 {
+                let id_op = Array2::<Complex64>::eye(usize::pow(2, extra_qarg2 as u32));
+                kron(&id_op, &first_mat)
+            } else {
+                first_mat
+            };
+            let op12 = compose(&first_mat.view(), &second_mat.view(), &second_qarg, false);
+            let op21 = compose(&first_mat.view(), &second_mat.view(), &second_qarg, true);
+            abs_diff_eq!(op12, op21, epsilon = 1e-8)
+        }
+    }
+
+    fn clear_cache(&mut self) {
+        self.cache.clear();
+        self.current_cache_entries = 0;
+        self._cache_miss = 0;
+        self._cache_hit = 0;
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn commutation_precheck(
+    op1: &CircuitInstruction,
+    qargs1: &[usize],
+    cargs1: &[usize],
+    op2: &CircuitInstruction,
+    qargs2: &[usize],
+    cargs2: &[usize],
+    max_num_qubits: u32,
+) -> Option<bool> {
```

Comment:
Again, consider using `u32` to represent qubit indices.
```suggestion
fn commutation_precheck(
    op1: &CircuitInstruction,
    qargs1: &[u32],
    cargs1: &[u32],
    op2: &CircuitInstruction,
    qargs2: &[u32],
    cargs2: &[u32],
    max_num_qubits: u32,
) -> Option<bool> {
```